fabric-gamerule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/rule/EnumRule.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.gamerule.v1.rule;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.world.GameRules;
+
+import net.fabricmc.fabric.api.gamerule.v1.GameRuleRegistry;
+
+public class EnumRule<E extends Enum<E>> extends LiteralRule<EnumRule<E>> {
+	private static final Logger LOGGER = LogManager.getLogger(GameRuleRegistry.class);
+
+	private final Class<E> classType;
+	protected final Collection<E> supportedValues;
+	protected E value;
+
+	@Deprecated
+	public EnumRule(GameRules.RuleType<EnumRule<E>> type, E value, E[] supportedValues) {
+		this(type, value, Arrays.asList(supportedValues));
+	}
+
+	@Deprecated
+	public EnumRule(GameRules.RuleType<EnumRule<E>> type, E value, Collection<E> supportedValues) {
+		super(type);
+		this.classType = value.getDeclaringClass();
+		this.value = value;
+		this.supportedValues = Collections.unmodifiableCollection(supportedValues);
+	}
+
+	@Override
+	protected void deserialize(String value) {
+		try {
+			final E deserialized = Enum.valueOf(this.classType, value);
+
+			if (!this.supports(deserialized)) {
+				LOGGER.warn("Failed to parse rule of value {} for rule of type {}. Since the value {}, is unsupported.", value, this.classType, value);
+			}
+
+			this.set(deserialized, null);
+		} catch (IllegalArgumentException e) {
+			LOGGER.warn("Failed to parse rule of value {} for rule of type {}", value, this.classType);
+		}
+	}
+
+	@Override
+	public String serialize() {
+		return this.value.name();
+	}
+
+	@Override
+	public int getCommandResult() {
+		return 0;
+	}
+
+	@Override
+	protected EnumRule<E> getThis() {
+		return this;
+	}
+
+	public Class<E> getEnumClass() {
+		return this.classType;
+	}
+
+	@Override
+	public String toString() {
+		return this.value.toString();
+	}
+
+	@Override
+	protected EnumRule<E> copy() {
+		return new EnumRule<E>(this.type, this.value, this.supportedValues);
+	}
+
+	@Override
+	public void setValue(EnumRule<E> rule, MinecraftServer minecraftServer) {
+		this.value = rule.value;
+		this.changed(minecraftServer);
+	}
+
+	public E get() {
+		return this.value;
+	}
+
+	public E cycle(E start) {
```

Comment:
What is the purpose of this? Why does it need to be part of this object?

Reply:
Used to cycle the value within the list of supported values on the gamerules screen.